### ISSUE TITLE
Make available in maven repositoriy

### ISSUE THREAD

**Issue by langurmonkey** [Starlink/starjava#54]
AFAIK, the only available version of STIL in a public mvn repo is `3.1` [here](https://mvnrepository.com/artifact/uk.ac.bristol.star/stil). If your project depends on a newer version, you need to download the jar manually or set up a script especially for pulling this one dependency. I would ask to publish recent versions in a public repository so that it can be pulled like any other dep.

**Comment by pahjbo** [comment]
I agree that it would be really good to have the starlink libraries published to a public repository - I have a fork https://github.com/Javastro/starjava that is solely there to create a maven based build of starjava, but I cannot publish the artefacts as I am not the overall owner. There are some subtleties with dependencies (e.g. depending on a patched version of a well known library) that would also have to be solved before being able to reliably publish.

**Comment by mbtaylor** [comment]
You're right, this should be done, it's mostly my laziness/ignorance about Maven that has prevented it so far.

I don't know how that v3.1 version got there, I don't *think* it was me that published it.  However, a more recent version [stil-io 3.3.2](https://bintray.com/opencadc/software/stil-io) is available, that was kindly published by @brianmajor.   This is missing some of the dependencies that you probably don't need.

I am currently working on a major upgrade (STIL 4.0) which among other things will remove some dependencies, so it should make publishing to a repo more straightforward.  When that's done, I will try to make publishing to a maven repo part of the release process. Since that's something I'm currently ignorant about, if you have any non-obvious tips about how I ought to do that, feel free to pass them on.

**Comment by langurmonkey** [comment]
Looks like the [repository](http://nexus.vamdc.org/nexus/content/repositories/releases/) hosting [jsamp](https://mvnrepository.com/artifact/org.astrogrid/jsamp/1.3.7) went offline, so I'm currently hosting jars for both stil and jsamp in my own repo.
Maybe [this](https://maven.apache.org/repository/guide-central-repository-upload.html) would help?

**Comment by mbtaylor** [comment]
Apologies @langurmonkey (and others) for slow progress on this, but I am working on it - I secured access to the sonatype uk.ac.starlink groupId within the last hour!  I'm hoping I can have STIL in the central repo in the next 2-3 weeks.

**Comment by mbtaylor** [comment]
I don't know about JSAMP, I wasn't aware of the repository that was in, or that it had gone down.  If I can host it under the same groupId (uk.ac.starlink) I'll probably try to do that, but I haven't got my head around the rules for what's permitted yet.

**Comment by pahjbo** [comment]
You can do what you want under the groupId now that you have ownership!

Just to point out again, I have already done a fair chunk of this work in https://github.com/Javastro/starjava (though it needs an update to get bring it up to date with your upstream). Even if you decide not to use maven, the POMs would be useful.

You also might want to consider how you number your intermediate versions (e.g. 3.4+), as they are not very maven friendly - though it might not matter if you never publish these.

Another thing that came out of mavenising was that there were some circular dependencies between the modules as currently split (I think only in the tests though - if I remember correctly).

**Comment by mbtaylor** [comment]
Thanks for that reminder Paul.  My current intention is to publish STIL as a more-or-less monolithic component rather than having the maven modules mapped 1:1 with starjava packages - does that make sense?  If so I think it should take care of the circular dependencies (as you say, I think they are only in the unit tests).  A few other components like JCDF (on which STIL would depend) and JSAMP (on which it wouldn't) probably make sense to publish as well.  At STIL 4.0 there are some simplifications of the dependencies, since I no longer have any crossmatching functionality as part of STIL (it's farmed out to starjava package TJOIN rather then being in TABLE).

But I'm not currently planning to publish the applications TOPCAT and STILTS or other things that they depend on unless there's some pressing reason to do so.  Similarly I wouldn't plan to publish intermediate releases.

I'm not especially enthusiastic about grappling with maven, so if I can steal or build on your work, or if you can set up the publication and I can stamp it with the groupId or something, it's fine with me.  What do you think?

**Comment by pahjbo** [comment]
I do not think that there is a particular advantage to publishing STIL as a monolithic component - once you have gone the maven route, all the dependencies are captured, so it does not matter that there are lots. In fact, you would find it harder to  do the monolithic approach than the 1:1 modules approach - unless you went for the monolithic approach with all 3rd party dependencies bundled too - but that would be severely sub-optimal for others to reuse that might depend on different versions of those dependencies.

I went for the 1:1 approach in what I did, and so you could easily adopt that - It is currently a bit behind at stil 3.3-3, but I could update to latest upstream, and do some more updates to the POM to follow best practices that I got halfway through, and at the same time add the central publishing stuff - I could then do a PR for you to merge. I am having a few days off, but could do the above by middle of next week probably.

**Comment by mbtaylor** [comment]
All right thanks @pahjbo let me have a think about it and I'll let you know if I want some help.

**Comment by mbtaylor** [comment]
@langurmonkey, the [VAMDC repository](http://nexus.vamdc.org/nexus/content/repositories/releases/) hosting JSAMP seems to be back up at the moment.  Is that OK or do you still think that JSAMP needs rehosting?

**Comment by langurmonkey** [comment]
I won't be using VAMDC anymore, as it still uses unencrypted, insecure HTTP. I still think it is worth it to try and get the builds into maven central or something like that. Aren't there any other users requesting this? I'm certainly not the only user of STIL and/or JSAMP willing to automatically pull the dependency from a repository, right?

**Comment by mbtaylor** [comment]
I have deployed STIL v4.0-2, as a more-or-less monolithic component, to the central repository under groupId [uk.ac.starlink](https://search.maven.org/search?q=g:uk.ac.starlink).  The STIL package is actually built outside of the starjava repository (under makefile control) so it seemed like an appropriate way to do it rather than working with multiple `pom.xml` files and version identifiers for the various internal sub-components.  The third-party dependencies (nom.tam.fits, snakeyaml, json.org, ant for which I need only a few classes related to bzip2 decompression, and a couple of other externally deployed libraries of mine) are referenced in what I believe is a sensible way; STIL itself doesn't have that many external dependencies since v4.0, as I've excised the crossmatching functionality which brought in a number of additional libraries.

As I've already admitted, I don't have much experience with maven and related infrastructure, so I'd welcome any comments about what I've got wrong.  I can modify the way it's done in future releases if necessary.  At present I'm not planning to make intermediate releases in this way, just to issue new artifacts for STIL public numbered releases.

I'll look at getting JSAMP into the central repository as well.

**Comment by langurmonkey** [comment]
Thanks Mark! Added the dependency to my project and so far everything looks good on my end. The transtitive dependencies get pulled correctly and everything works as expected.

**Comment by vforchi** [comment]
As already discussed, the monolithic package contains a little too much for me, and pulls dependencies that I don't really need.
A possible way to address this is to split stil into two or more packages, e.g. stil-io, stil-table, stil-core... I can't really suggest the naming, because I don't know the interdependencies of the modules.
For my particular project I need only fits/table/votable/util, and the only external dependency that they require is `nom.tam.fits`, I think.

**Comment by mbtaylor** [comment]
@vforchi I see your point, but this kind of dependency proliferation seems to be rather endemic in the maven world, and I wouldn't say that this STIL package is one of the worst offenders. 

Possibly organising the package along the lines that @pahjbo has suggested, which I think would complicate the packaging from my point of view, would make a la carte dependency picking possible.  But I don't know ... the runtime (though not build-time) dependencies of the `table` package really are all the other format-specific handler packages, in the sense that some calls to e.g. `StarTableFactory` will fail with a `NoClassDefFoundError` if those other packages are absent at runtime (though it's true that if you avoid those calls you can get away with omitting the unused handlers).  I could do something to make such failures more graceful, but again it's more work.

Anyway: my current plan is to stick at about this level of granularity, and if you want something more custom you'll have to do it by hand.  If I hear strong arguments otherwise I might reconsider.  One improvement on the way though: I probably will try to rework the provisioning of the bzip2 code to avoid pulling in ant classes for future releases, so some of these dependencies (`ant`, `ant-launcher`, `hamcrest-core`) should go away in future.

**Comment by langurmonkey** [comment]
> One improvement on the way though: I probably will try to rework the provisioning of the bzip2 code to avoid pulling in ant classes for future releases, so some of these dependencies (ant, ant-launcher, hamcrest-core) should go away in future.

That would be very welcome indeed. I also noticed an increase in the size of my libraries due to these new deps. `ant` alone is over 1 MB.

**Comment by vforchi** [comment]
In my opinion inclusion of runtime dependencies should be the responsibility of the end user.
E.g. if you used `slf4j`, you shouldn't include a backend, the choice should be left to the final user.
Currently I am importing `stil` by hand like this (`gradle`):
```kotlin
    implementation(files("libs/stil-4.0.1/fits.jar"))
    implementation("gov.nasa.gsfc.heasarc:nom-tam-fits:1.15.2")
    implementation(files("libs/stil-4.0.1/table.jar"))
    implementation(files("libs/stil-4.0.1/votable.jar"))
    implementation(files("libs/stil-4.0.1/util.jar"))
```
Since I am reluctant in adding unnecessary dependencies I will stick to this until the next non-minor release, and then I will reconsider it.

**Comment by mbtaylor** [comment]
@langurmonkey, JSAMP 1.3.7 is now available under groupId `uk.ac.starlink`.  Let me know if it looks OK.

**Comment by langurmonkey** [comment]
@mbtaylor everything loos fine with `jsamp` as well, thanks!

**Comment by mbtaylor** [comment]
I'm going to close this issue, since STIL (and JSAMP) are now available from the central repository, though perhaps not 100% to everybody's satisfaction.  If there are more comments it could be reopened or another issue started.

**Comment by mbtaylor** [comment]
Postscript: @langurmonkey I have just released STIL 4.0-3, which avoids the ant dependency so is a bit more compact.

**Comment by langurmonkey** [comment]
@mbtaylor awesome, thanks!
